array_api_tests/test_special_cases.py

```diff
@@ -0,0 +1,646 @@
+from . import shape_helpers as sh
+from . import xps
+from ._array_module import mod as xp
+from .stubs import category_to_funcs
+
+pytestmark = pytest.mark.ci
+
+UnaryCheck = Callable[[float], bool]
+BinaryCheck = Callable[[float, float], bool]
+
+
+def make_eq(v: float) -> UnaryCheck:
+    if math.isnan(v):
+        return math.isnan
+    if v == 0:
+        if ph.is_pos_zero(v):
+            return ph.is_pos_zero
+        else:
+            return ph.is_neg_zero
+
+    def eq(i: float) -> bool:
+        return i == v
+
+    return eq
+
+
+def make_neq(v: float) -> UnaryCheck:
+    eq = make_eq(v)
+
+    def neq(i: float) -> bool:
+        return not eq(i)
+
+    return neq
+
+
+def make_rough_eq(v: float) -> UnaryCheck:
+    assert math.isfinite(v)  # sanity check
+
+    def rough_eq(i: float) -> bool:
+        return math.isclose(i, v, abs_tol=0.01)
+
+    return rough_eq
+
+
+def make_gt(v: float) -> UnaryCheck:
+    assert not math.isnan(v)  # sanity check
+
+    def gt(i: float) -> bool:
+        return i > v
+
+    return gt
+
+
+def make_lt(v: float) -> UnaryCheck:
+    assert not math.isnan(v)  # sanity check
+
+    def lt(i: float) -> bool:
+        return i < v
+
+    return lt
+
+
+def make_or(cond1: UnaryCheck, cond2: UnaryCheck) -> UnaryCheck:
+    def or_(i: float) -> bool:
+        return cond1(i) or cond2(i)
+
+    return or_
+
+
+def make_and(cond1: UnaryCheck, cond2: UnaryCheck) -> UnaryCheck:
+    def and_(i: float) -> bool:
+        return cond1(i) or cond2(i)
+
+    return and_
+
+
+def notify_cond(cond: UnaryCheck) -> UnaryCheck:
+    def not_cond(i: float) -> bool:
+        return not cond(i)
+
+    return not_cond
+
+
+def absify_cond(cond: UnaryCheck) -> UnaryCheck:
+    def abs_cond(i: float) -> bool:
+        return cond(abs(i))
+
+    return abs_cond
+
+
+repr_to_value = {
+    "NaN": float("nan"),
+    "infinity": float("inf"),
+    "0": 0.0,
+    "1": 1.0,
+}
+r_value = re.compile(r"([+-]?)(.+)")
+r_pi = re.compile(r"(\d?)π(?:/(\d))?")
+
+
+@dataclass
+class ValueParseError(ValueError):
+    value: str
+
+
+def parse_value(value_str: str) -> float:
+    m = r_value.match(value_str)
+    if m is None:
+        raise ValueParseError(value_str)
+    if pi_m := r_pi.match(m.group(2)):
+        value = math.pi
+        if numerator := pi_m.group(1):
+            value *= int(numerator)
+        if denominator := pi_m.group(2):
+            value /= int(denominator)
+    else:
+        value = repr_to_value[m.group(2)]
+    if sign := m.group(1):
+        if sign == "-":
+            value *= -1
+    return value
+
+
+r_code = re.compile(r"``([^\s]+)``")
+r_approx_value = re.compile(
+    rf"an implementation-dependent approximation to {r_code.pattern}"
+)
+
+
+def parse_inline_code(inline_code: str) -> float:
+    if m := r_code.match(inline_code):
+        return parse_value(m.group(1))
+    else:
+        raise ValueParseError(inline_code)
+
+
+r_not = re.compile("not (?:equal to )?(.+)")
```

Comment:
It's good to always use make regular expressions raw strings, even if they don't have backslashes. You don't want to add a backslash escape later only to not realize the string isn't raw.